primitives/beefy/src/commitment.rs

```diff
@@ -183,28 +183,28 @@ struct CompactSignedCommitment<TCommitment> {
 	/// Note that in order to associate a `Signature` from this `Vec` with a validator, one needs
 	/// to look at the `signatures_from` bitfield, since some validators might have not produced a
 	/// signature.
-	signatures_compact: Vec<Signature>,
+	signatures_compact: Vec<TSignature>,
 }
 
-impl<'a, TBlockNumber> CompactSignedCommitment<&'a Commitment<TBlockNumber>> {
+impl<'a, TBlockNumber: Clone, TSignature> CompactSignedCommitment<TBlockNumber, &'a TSignature> {
 	/// Packs a `SignedCommitment` into the compressed `CompactSignedCommitment` format for
 	/// efficient network transport.
-	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber>) -> Self {
+	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber, TSignature>) -> Self {
 		let SignedCommitment { commitment, signatures } = signed_commitment;
 		let validator_set_len = signatures.len() as u32;
 		let mut signatures_from: BitField = vec![];
-		let mut signatures_compact: Vec<Signature> = vec![];
+		let mut signatures_compact: Vec<&'a TSignature> = vec![];
+
+		let mut bits: Vec<u8> =
```

Comment:
can't we do this and the next block in one iteration?

Reply:
Yes, but actually I think I'd rather use iterators for both:
```rust
let mut bits = signatures.iter().map(|x| if x.is_some() { 1 } else { 0 }).collect();
let mut signatures_compact = signatures.iter().filter_map(|x| x).collect();
```

wdyt?